### PR TITLE
DOP-3993: Allow card group to link data for IA entry

### PR DESCRIPTION
**Pull request by rayangler**
### Ticket

DOP-3993

### Notes

* Adds an optional id to IA entries. This ID can be referenced by a `card-group` directive to link card data with an IA entry. This data will be usable in the side nav.
  * The concept of linking data this way might be weird... This is for a (probably) one-off instance for the new homepage, but hopefully this is okay. I'm open to suggestions!
* Linked data will be present in the metadata to allow different pages with similar IA entries to render the linked data in multiple places.
* Relevant [frontend PR](https://github.com/mongodb/snooty/pull/890)
* Relevant [rST commit](https://github.com/mongodb/docs-landing/pull/132/commits/d157270c47a843add7f83e529357be7bcded8825)

### CODE REVIEW

File: snooty/test_postprocess.py [hunk 1]
```diff
@@ -155,6 +156,93 @@ def test_ia() -> None:
...
+
+        diagnostics = result.diagnostics[FileId("page1.txt")]
+        assert len(diagnostics) == 1
+        assert isinstance(diagnostics[0], InvalidIALinkedData)
+
+        # Ensure the one IA entry that has linked data contains said linked data.
+        metadata = cast(Dict[str, Any], result.metadata)
+        ia_tree = metadata["iatree"]
+        root_children = ia_tree["children"]
+        for entry in root_children:
```
Comment (branberry) [inline snooty/test_postprocess.py:237]: Thank you for the helpful tests! These helped solidify my understanding of the changes made to the parser.

**Review by branberry — APPROVED** [review]
LGTM! Just had one tiny question, but everything makes sense to me.

File: snooty/postprocess.py [hunk 2]
```diff
@@ -1132,21 +1136,64 @@ def serialize(self) -> n.SerializedNode:
...
+                )
+                continue
+
+            self.entry_ids[entry_id].append(card.options)
 
     def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
-        if (
-            not isinstance(node, n.Directive)
-            or not node.name == "ia"
-            or not node.domain == ""
```
[the remark this answers is not in the document]
Comment (i80and) [reply snooty/postprocess.py:1146]: We definitely do want to check the domain

File: snooty/postprocess.py [hunk 3]
```diff
@@ -1132,21 +1136,64 @@ def serialize(self) -> n.SerializedNode:
...
+        if not isinstance(node, n.Directive) or not (
+            node.name == "ia" or node.name == "card-group"
         ):
             return
 
+        # A card-group directive can have data linked to a particular IA entry
+        # for the side nav
+        if node.name == "card-group":
+            entry_id = node.options.get("ia-entry-id", None)
```
Comment (i80and) [inline snooty/postprocess.py:1184]: `None` is actually implied in `.get()`

File: snooty/postprocess.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rayangler) [reply snooty/postprocess.py:1146]: Yup, this should not have been removed. I have re-added "" and "mongodb" (for `card-group`) as valid domains

File: snooty/postprocess.py [hunk 4]
```diff
@@ -1132,21 +1136,64 @@ def serialize(self) -> n.SerializedNode:
...
+                result["id"] = self.entry_id
+                if self.entry_id in entry_ids:
+                    result["linked_data"] = entry_ids[self.entry_id]
 
             return result
 
     def __init__(self, context: Context) -> None:
         super().__init__(context)
         self.ia: List[IAHandler.IAData] = []
+        self.entry_ids: Dict[str, List[Dict[str, str]]] = {}
```
[the remark this answers is not in the document]
Comment (rayangler) [reply snooty/postprocess.py:1149]: Closing this thread as it was discussed on Slack

**Review by i80and — APPROVED** [review]

**Pull request merged**